watchtower-plugin/src/convert.rs

```diff
@@ -0,0 +1,435 @@
+        }
+    }
+}
+
+/// Parameters related to the `registertower` command.
+#[derive(Debug, Serialize)]
+pub struct RegisterParams {
+    pub tower_id: TowerId,
+    pub host: Option<String>,
+    pub port: Option<u16>,
+}
+
+impl RegisterParams {
+    fn new(tower_id: &str, host: Option<&str>, port: Option<u64>) -> Result<Self, RegisterError> {
+        let mut params = RegisterParams::from_id(tower_id)?;
+
+        if host.is_some() {
+            params = params.with_host(host.unwrap())?
+        }
+
+        if port.is_some() {
+            params = params.with_port(port.unwrap())?
+        }
+
+        Ok(params)
+    }
+
+    fn from_id(tower_id: &str) -> Result<Self, RegisterError> {
+        Ok(Self {
+            tower_id: TowerId::from_str(tower_id)
+                .map_err(|_| RegisterError::InvalidId("Invalid tower id".into()))?,
+            host: None,
+            port: None,
+        })
+    }
+
+    fn with_host(self, host: &str) -> Result<Self, RegisterError> {
+        if host.is_empty() {
+            Err(RegisterError::InvalidHost("hostname is empty".into()))
+        } else if host.contains(' ') {
+            Err(RegisterError::InvalidHost(
+                "hostname contains white spaces".into(),
+            ))
+        } else {
+            Ok(Self {
+                host: Some(String::from(host)),
+                ..self
+            })
+        }
+    }
+
+    fn with_port(self, port: u64) -> Result<Self, RegisterError> {
+        if port > u16::MAX as u64 {
+            Err(RegisterError::InvalidPort(format!(
+                "port must be a 16-byte integer. Received: {}",
+                port
+            )))
+        } else {
+            Ok(Self {
+                port: Some(port as u16),
+                ..self
+            })
+        }
+    }
+}
+
+impl TryFrom<serde_json::Value> for RegisterParams {
+    type Error = RegisterError;
+
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        match value {
+            serde_json::Value::String(s) => {
+                let s = s.trim();
+                let mut v = s.split('@');
+                let tower_id = v.next().unwrap();
+
+                match v.next() {
+                    Some(x) => {
+                        let mut v = x.split(':');
+                        let host = v.next();
+                        let port = if let Some(p) = v.next() {
+                            p.parse()
+                                .map(Some)
+                                .map_err(|_| RegisterError::InvalidPort(format!("Port is not a number: {}", p)))?
+                        } else {
+                            None
+                        };
+
+                        RegisterParams::new(tower_id, host, port)
+                    }
+                    None => RegisterParams::from_id(tower_id),
+                }
+            }
+            serde_json::Value::Array(mut a) => {
+            let param_count = a.len();
```

Comment:
This line is missing an indentation. Weirdly though, `cargo fmt` doesn't fix this.